clash-lib/src/Clash/Netlist/Types.hs

```diff
@@ -81,23 +95,170 @@ data TopEntityT = TopEntityT
   -- ^ (Maybe) a test bench associated with the topentity
   } deriving (Generic, Show)
 
+-- | Same as "TopEntity", but with all port names that end up in HDL specified
+data ExpandedTopEntity a = ExpandedTopEntity
+  { et_inputs :: [Maybe (ExpandedPortName a)]
+  -- ^ Inputs with fully expanded port names. /Nothing/ if port is void.
+  , et_output :: Maybe (ExpandedPortName a)
+  -- ^ Output with fully expanded port names. /Nothing/ if port is void or
+  -- BiDirectionalOut.
+  } deriving (Show, Functor, Foldable, Traversable)
+
+-- | See "ExpandedTopEntity"
+data ExpandedPortName a
+  -- | Same as "PortName", but fully expanded
+  = ExpandedPortName HWType a
+
+  -- | Same as "PortProduct", but fully expanded
+  | ExpandedPortProduct
+      Text
+      -- FIELD Name hint. Can be used to create intermediate signal names.
+      HWType
+      -- FIELD Type of product
+      [ExpandedPortName a]
+      -- FIELD Product fields
+  deriving (Show, Functor, Foldable, Traversable)
+
 -- | Monad that caches generated components (StateT) and remembers hidden inputs
 -- of components that are being generated (WriterT)
 newtype NetlistMonad a =
-  NetlistMonad { runNetlist :: StateT NetlistState (ReaderT NetlistEnv IO) a }
+  NetlistMonad { runNetlist :: Strict.StateT NetlistState (ReaderT NetlistEnv IO) a }
   deriving newtype (Functor, Monad, Applicative, MonadReader NetlistEnv,
-                    MonadState NetlistState, MonadIO, MonadFail)
+                    Strict.MonadState NetlistState, Strict.MonadIO, MonadFail)
 
 type HWMap = HashMap Type (Either String FilteredHWType)
 
+-- | See 'is_freshCache'
+type FreshCache = HashMap Text (IntMap Word)
```

Comment:
This can be simplified: we can just store a single extension here and store additional extensions in the basename of the identifier. This would make the fresh cache just `HashMap Text Word`.

Reply:
Welp, this cannot be simplified. It's really hard to set it up such that you recognize that `foo` + extension `0` is the same as  `foo_0`, without blowing up the complexity class.